util/topgen.py

```diff
@@ -663,32 +663,53 @@ def generate_top_ral(top: Dict[str, object],
     # Collect up the memories to add
     mems = []
     for item in list(top.get("memory", [])):
-        byte_write = ('byte_write' in item and
-                      item["byte_write"].lower() == "true")
-        data_intg_passthru = ('data_intg_passthru' in item and
-                              item["data_intg_passthru"].lower() == "true")
-        size_in_bytes = int(item['size'], 0)
-        num_regs = size_in_bytes // addrsep
-        swaccess = access.SWAccess('top-level memory',
-                                   item.get('swaccess', 'rw'))
-
-        mems.append(window.Window(name=item['name'],
-                                  desc='(generated from top-level)',
-                                  unusual=False,
-                                  byte_write=byte_write,
-                                  data_intg_passthru=data_intg_passthru,
-                                  validbits=regwidth,
-                                  items=num_regs,
-                                  size_in_bytes=size_in_bytes,
-                                  offset=int(item["base_addr"], 0),
-                                  swaccess=swaccess))
+        offset = int(item['base_addr'], 0)
+        mems.append(create_mem(item, item['name'], addrsep, regwidth, offset))
+
+    # if memory is defined in top['module'], use it to replace the mem in the block
+    for module in top['module']:
+        if "memory" in module:
+            block = name_to_block[module['type']]
+            for mem_name, item in module['memory'].items():
+                assert block.reg_blocks[mem_name]
+                assert len(block.reg_blocks[mem_name].windows) == 1
+
+                if 'base_addr' in item:
+                    offset = int(item['base_addr'], 0)
+                else:
+                    offset = 0
+
+                block.reg_blocks[mem_name].windows[0] = create_mem(
+                    item, mem_name, addrsep, regwidth, offset)
```

Comment:
This feels wrong to me. Here, `block` is an `IpBlock` object which gives the type of the thing that's being instantiated. We're taking information from a specific instance (found via `module['memory']`) and writing it back to the type (overwriting what was there already), which is the wrong way around.

I suspect bad things will happen if there are multiple instances of the block instantiated in different ways but, more importantly, this doesn't really respect the data model.

If there's no way to make this work without altering regtool's data (which I think might be true), maybe the right thing is to make a copy of the IpBlock object, alter that copy, update module['type'] to some "derived" type and finally write the copy to `name_to_block` under the derived type name.

Or maybe there's a simpler approach? But I definitely don't think we should be altering the underlying types after they've been constructed.

Reply:
Good point. Let me make the change like this
assume we have 3 RAM block, and want to override 2 of them to 2 different types

```
// initial database:
name_to_block["RAM"] = ram_type
inst_to_block["RAM1"] = "RAM"
inst_to_block["RAM2"] = "RAM"
inst_to_block["RAM3"] = "RAM"

// updated database:
name_to_block["RAM"] = ram_type
name_to_block["RAM_NEW1"] = ram_type_new1
name_to_block["RAM_NEW2"] = ram_type_type_new2
inst_to_block["RAM1"] = "RAM"  // RAM1 still uses original type
inst_to_block["RAM2"] = "RAM_NEW1"
inst_to_block["RAM3"] = "RAM_NEW2"
```

This structure makes more sense. WDYT? @rswarbrick

Reply:
That looks like a good plan to me. (I'd forgotten what `name_to_block` was keyed by: your suggestion looks perfect!)